subsys/bluetooth/services/fast_pair/fp_advertising.c

```diff
@@ -71,20 +104,61 @@ size_t bt_fast_pair_adv_data_size(enum bt_fast_pair_adv_mode fp_adv_mode)
 
 	res += sizeof(fast_pair_uuid);
 
-	if (fp_adv_mode == BT_FAST_PAIR_ADV_MODE_DISCOVERABLE) {
+	if (fp_adv_config.adv_mode == BT_FAST_PAIR_ADV_MODE_DISCOVERABLE) {
 		res += bt_fast_pair_adv_data_size_discoverable();
 	} else {
-		res += bt_fast_pair_adv_data_size_non_discoverable(account_key_cnt);
+		res += bt_fast_pair_adv_data_size_non_discoverable(account_key_cnt, fp_adv_config);
 	}
 
 	return res;
 }
 
+static void fp_adv_data_fill_battery_info(uint8_t *battery_info,
+					  enum fp_field_type battery_data_type)
+{
+	struct net_buf_simple nb;
+	struct bt_fast_pair_battery_data battery_data = fp_battery_get_battery_data();
+	static const size_t battery_data_size = FP_CRYPTO_BATTERY_INFO_LEN - FIELD_LEN_TYPE_SIZE;
+
+	net_buf_simple_init_with_data(&nb, battery_info, FP_CRYPTO_BATTERY_INFO_LEN);
+	net_buf_simple_reset(&nb);
+
+	BUILD_ASSERT(sizeof(uint8_t) == FIELD_LEN_TYPE_SIZE);
+	BUILD_ASSERT(battery_data_size <= BIT_MASK(LEN_BITS));
+	BUILD_ASSERT(battery_data_size == BT_FAST_PAIR_BATTERY_VALUE_IDX_COUNT);
+	net_buf_simple_add_u8(&nb, ENCODE_FIELD_LEN_TYPE(battery_data_size, battery_data_type));
+
+	BUILD_ASSERT(sizeof(uint8_t) == BATTERY_VALUE_SIZE);
+
+	for (int8_t i = 0; i < battery_data_size; i++) {
+		__ASSERT_NO_MSG(battery_data.battery_values[i].level <=
+				BIT_MASK(BATTERY_LEVEL_BITS));
+		net_buf_simple_add_u8(&nb, ENCODE_FIELD_BATTERY_STATUS_LEVEL(
+							battery_data.battery_values[i].charging,
+							battery_data.battery_values[i].level));
+	}
+}
+
 static int fp_adv_data_fill_non_discoverable(struct net_buf_simple *buf, size_t account_key_cnt,
-					     enum fp_field_type ak_filter_type)
+					     enum fp_field_type ak_filter_type,
+					     struct bt_fast_pair_adv_config fp_adv_config)
```

Comment:
I would pass here only the selected battery info mode to avoid passing unused information.